File: frontend/src/api-client/todos.ts
const API_BASE = import.meta.env.VITE_API_BASE_URL ?? '';

export type Todo = {
  id: string;
  title: string;
  completed: boolean;
  created_at: string;
  updated_at: string;
};

type TodoCreate = {
  title: string;
};

type TodoUpdate = {
  completed: boolean;
};

export class ApiError extends Error {
  status: number;

  constructor(status: number, message: string) {
    super(message);
    this.status = status;
  }
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...(init?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const message = await response.text();
    throw new ApiError(response.status, message || response.statusText);
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return (await response.json()) as T;
}

export function fetchTodos() {
  return request<Todo[]>('/api/todos');
}

export function createTodo(payload: TodoCreate) {
  return request<Todo>('/api/todos', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function updateTodo(id: string, payload: TodoUpdate) {
  return request<Todo>(`/api/todos/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

export function deleteTodo(id: string) {
  return request<void>(`/api/todos/${id}`, {
    method: 'DELETE',
  });
}

File: backend/app/routers/todos.py
# API CONTRACT
# GET  /api/todos
#   response: [
#     {"id": int, "title": str, "completed": bool, "created_at": str, "updated_at": str}
#   ]
# POST /api/todos
#   request:  {"title": str}
#   response: {"id": int, "title": str, "completed": bool, "created_at": str, "updated_at": str}
# PATCH /api/todos/{id}
#   request:  {"completed": bool}
#   response: {"id": int, "title": str, "completed": bool, "created_at": str, "updated_at": str}
# DELETE /api/todos/{id}
#   response: {"ok": true}

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Path, status
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.todo import Todo
from app.schemas.todo import TodoCreate, TodoOut, TodoUpdate

router = APIRouter(prefix="/api/todos", tags=["todos"])


@router.get("", response_model=list[TodoOut])
async def list_todos(db: Session = Depends(get_db)) -> list[Todo]:
    return db.query(Todo).order_by(Todo.id.asc()).all()


@router.post("", response_model=TodoOut, status_code=status.HTTP_201_CREATED)
async def create_todo(body: TodoCreate, db: Session = Depends(get_db)) -> Todo:
    todo = Todo(title=body.title, completed=False)
    db.add(todo)
    try:
        db.flush()
        db.refresh(todo)
    except SQLAlchemyError as exc:
        raise HTTPException(status_code=500, detail={"error": "todo_create_failed", "message": "Failed to create todo"}) from exc
    return todo


@router.patch("/{todo_id}", response_model=TodoOut)
async def update_todo(todo_id: int = Path(ge=1), body: TodoUpdate | None = None, db: Session = Depends(get_db)) -> Todo:
    todo = db.get(Todo, todo_id)
    if todo is None:
        raise HTTPException(status_code=404, detail={"error": "todo_not_found", "message": "Todo not found"})
    assert body is not None
    todo.completed = body.completed
    try:
        db.flush()
        db.refresh(todo)
    except SQLAlchemyError as exc:
        raise HTTPException(status_code=500, detail={"error": "todo_update_failed", "message": "Failed to update todo"}) from exc
    return todo


@router.delete("/{todo_id}")
async def delete_todo(todo_id: int = Path(ge=1), db: Session = Depends(get_db)) -> dict[str, bool]:
    todo = db.get(Todo, todo_id)
    if todo is None:
        raise HTTPException(status_code=404, detail={"error": "todo_not_found", "message": "Todo not found"})
    try:
        db.delete(todo)
        db.flush()
    except SQLAlchemyError as exc:
        raise HTTPException(status_code=500, detail={"error": "todo_delete_failed", "message": "Failed to delete todo"}) from exc
    return {"ok": True}
